File: DevEnvironment/Base/src/main/java/base/core/UnixPath.java
package base.core;

import base.interfaces.IPathBuilder;
import java.util.Stack;

/**
 * The UnixPath class implements a unix path.
 */
public final class UnixPath extends AbstractPath {
    public static final char SEPARATOR = '/';
    public static final String ROOT_DIRECTORY = "/";
    public static final String DIRECTORY_SEPARATOR = String.valueOf(SEPARATOR);
    public static final String CURRENT_DIRECTORY = Paths.CURRENT_DIRECTORY;
    public static final String PARENT_DIRECTORY = Paths.PARENT_DIRECTORY;
    public static final String HOME_DIRECTORY = "~";

    /**
     * The UnixPath constructor.
     */
    public UnixPath(String path) {
        super(path);
    }

    /**
     * Gets the directory section of the path.
     */
    public static String getDirectory(String path) {
        return Paths.getDirectory(path, UnixPath.DIRECTORY_SEPARATOR);
    }

    /**
     * Makes an absolute path.
     */
    @Override
    protected String makeAbsolutePath(String path) {
        return UnixPath.makeAbsolute(path);
    }

    /**
     * Makes a canonical path.
     */
    @Override
    protected String makeCanonicalPath(String path) {
        return UnixPath.makeCanonical(path);
    }

    /**
     * Validates a path.
     */
    @Override
    protected void validatePath(String path) {
        UnixPath.validate(path);
    }

    /**
     * Validates a unix path.
     */
    public static void validate(String path) {
        Conditions.validateStringNotNullOrEmpty(
            path,
            "The unix path.");
    }

    /**
     * Makes an absolute path.
     */
    private static String makeAbsolute(String path) {
        String homeDirectory;
        String reminder;

        if (path.startsWith(UnixPath.HOME_DIRECTORY)) {
            homeDirectory = System.getProperty("user.home");
            reminder = path.substring(UnixPath.HOME_DIRECTORY.length());
        }
        else {
            homeDirectory = "";
            reminder = path;
        }

        String canonicalPath = makeCanonical(reminder);

        IPathBuilder pathBuilder = new UnixPathBuilder();

        if (!homeDirectory.isEmpty()) {
            pathBuilder.addComponent(homeDirectory);
        }

        pathBuilder.addComponent(canonicalPath);

        return pathBuilder.build();
    }

    /**
     * Makes a canonical path.
     */
    private static String makeCanonical(String path) {
        String[] components = path.split(UnixPath.DIRECTORY_SEPARATOR);

        Stack<String> stack = new Stack<>();

        for (String component : components) {
            if (component.isEmpty() || component.equals(UnixPath.CURRENT_DIRECTORY)) {
            }
            else if (component.equals(UnixPath.PARENT_DIRECTORY)) {
                if (!stack.empty()) {
                    stack.pop();
                }
            }
            else {
                stack.push(component);
            }
        }

        IPathBuilder pathBuilder = new UnixPathBuilder();

        for (String component : stack) {
            pathBuilder.addComponent(component);
        }

        return pathBuilder.build();
    }
}

File: SystemCore/Base/core/UnixPath.cpp
#include "UnixPath.h"
#include "PathBuilder.h"
#include "Strings.h"
#include "InvalidArgumentException.h"
#include "NotImplementedException.h"
#include <stack>

using namespace base;
using namespace base::environment::unix;
const std::wstring::value_type UnixPath::SEPARATOR = L'/';
const std::wstring UnixPath::DIRECTORY_SEPARATOR = std::wstring(1, SEPARATOR);
const std::wstring::value_type UnixPath::EXTENSION_SEPARATOR = L'.';
const std::wstring UnixPath::CURRENT_DIRECTORY = L".";
const std::wstring UnixPath::PARENT_DIRECTORY = L"..";
const std::wstring UnixPath::HOME_DIRECTORY = L"~";

/**
 * Creates a new unix path.
 */
IPathSharedPtr UnixPath::Make(const std::wstring& path)
{
    return std::make_shared<UnixPath>(path);
}

/**
 * The UnixPath constructor.
 */
UnixPath::UnixPath(const std::wstring& path) :
    AbstractPath(
        path,
        DIRECTORY_SEPARATOR,
        EXTENSION_SEPARATOR)
{
}

/**
 * The UnixPath destructor.
 */
UnixPath::~UnixPath()
{
}

/**
 * Gets seperator.
 */
std::wstring::value_type UnixPath::Seperator()
{
    return SEPARATOR;
}

/**
 * Gets directory seperator.
 */
std::wstring UnixPath::DirectorySeperator()
{
    return DIRECTORY_SEPARATOR;
}

/**
 * Gets current directory.
 */
std::wstring UnixPath::CurrentDirectory()
{
    return CURRENT_DIRECTORY;
}

/**
 * Gets parent directory.
 */
std::wstring UnixPath::ParentDirectory()
{
    return PARENT_DIRECTORY;
}

/**
 * Validates a path.
 */
void UnixPath::ValidatePath(const std::wstring& path) const
{
    if (path.empty()) {
        throw InvalidArgumentException(L"The input path can not be empty.");
    }
}

/**
 * Makes an absolute path.
 */
std::wstring UnixPath::MakeAbsolutePath(const std::wstring& path) const
{
    return MakeAbsolute(path);
}

/**
 * Makes a canonical path.
 */
std::wstring UnixPath::MakeCanonicalPath(const std::wstring& path) const
{
    return MakeCanonical(path);
}

/**
 * Makes an absolute path.
 */
std::wstring UnixPath::MakeAbsolute(const std::wstring& path)
{
    throw NotImplementedException(L"UnixPath::MakeAbsolute()");
}

/**
 * Makes a canonical path.
 */
std::wstring UnixPath::MakeCanonical(const std::wstring& path)
{
    //
    // Split tokens by directory seperator...
    //
    std::vector<std::wstring> tokens;
    Strings::SplitString(path, DirectorySeperator(), tokens);

    //
    // Process the tokens with a stack...
    //
    std::stack<std::wstring> stack;

    for (const std::wstring& token : tokens) {
        if (token == ParentDirectory())
        {
            stack.pop();
        }
        else if (!token.empty() && token != CurrentDirectory())
        {
            stack.push(token);
        }
    }

    //
    // Create the list of finalize component...
    //
    std::list<std::wstring> components;

    while (!stack.empty()) {
        std::wstring component = stack.top();
        stack.pop();
        components.push_front(component);
    }

    //
    // Assemble the path...
    //
    PathBuilder pathBuilder;

    for (const std::wstring& component : components) {
        pathBuilder.AddComponent(component);
    }

    return pathBuilder.Build();
}

/**
 * Checks whether a path is absolute.
 */
bool UnixPath::IsAbsolute(const std::wstring& path) const
{
    return !IsRelative(path);
}

/**
 * Checks whether a path is relative.
 */
bool UnixPath::IsRelative(const std::wstring& path) const
{
    throw NotImplementedException(L"UnixPath::IsRelative()");
}
